File: go/internal/schema/schema.go
// Package schema defines word and dictionary data structures for ditong.
package schema

import (
	"encoding/json"
	"os"
	"path/filepath"
	"sort"
	"time"
)

// WordSource tracks where a word came from.
type WordSource struct {
	DictName     string `json:"dict_name"`
	DictFilepath string `json:"dict_filepath"`
	Language     string `json:"language"`
	OriginalForm string `json:"original_form"`
	LineNumber   *int   `json:"line_number,omitempty"`
	Category     string `json:"category"`
}

// Word represents a normalized word with full metadata.
type Word struct {
	Normalized      string            `json:"normalized"`
	Length          int               `json:"length"`
	WordType        string            `json:"type"`
	Sources         []WordSource      `json:"sources"`
	Categories      map[string]bool   `json:"-"` // Internal set
	Languages       map[string]bool   `json:"-"` // Internal set
	Tags            map[string]bool   `json:"-"` // Internal set
	IPA             string            `json:"ipa,omitempty"`
	SynthesisGroups map[string]bool   `json:"-"` // Internal set
}

// NewWord creates a new Word with initialized sets.
func NewWord(normalized string, length int, wordType string) *Word {
	return &Word{
		Normalized:      normalized,
		Length:          length,
		WordType:        wordType,
		Sources:         []WordSource{},
		Categories:      make(map[string]bool),
		Languages:       make(map[string]bool),
		Tags:            make(map[string]bool),
		SynthesisGroups: make(map[string]bool),
	}
}

// AddSource adds a source and updates derived fields.
func (w *Word) AddSource(source WordSource) {
	w.Sources = append(w.Sources, source)
	w.Categories[source.Category] = true
	w.Languages[source.Language] = true
}

// GetSourceDicts returns list of source dictionary names.
func (w *Word) GetSourceDicts() []string {
	dicts := make([]string, len(w.Sources))
	for i, s := range w.Sources {
		dicts[i] = s.DictName
	}
	return dicts
}

// MatchesFilter checks if word matches filter criteria.
func (w *Word) MatchesFilter(
	includeCategories, excludeCategories map[string]bool,
	includeLanguages map[string]bool,
	minLength, maxLength int,
) bool {
	if minLength > 0 && w.Length < minLength {
		return false
	}
	if maxLength > 0 && w.Length > maxLength {
		return false
	}

	if len(includeLanguages) > 0 {
		found := false
		for lang := range w.Languages {
			if includeLanguages[lang] {
				found = true
				break
			}
		}
		if !found {
			return false
		}
	}

	if len(includeCategories) > 0 {
		found := false
		for cat := range w.Categories {
			if includeCategories[cat] {
				found = true
				break
			}
		}
		if !found {
			return false
		}
	}

	if len(excludeCategories) > 0 {
		for cat := range w.Categories {
			if excludeCategories[cat] {
				return false
			}
		}
	}

	return true
}

// MarshalJSON implements custom JSON marshaling.
func (w *Word) MarshalJSON() ([]byte, error) {
	type Alias Word

	categories := make([]string, 0, len(w.Categories))
	for k := range w.Categories {
		categories = append(categories, k)
	}
	sort.Strings(categories)

	languages := make([]string, 0, len(w.Languages))
	for k := range w.Languages {
		languages = append(languages, k)
	}
	sort.Strings(languages)

	tags := make([]string, 0, len(w.Tags))
	for k := range w.Tags {
		tags = append(tags, k)
	}
	sort.Strings(tags)

	synthGroups := make([]string, 0, len(w.SynthesisGroups))
	for k := range w.SynthesisGroups {
		synthGroups = append(synthGroups, k)
	}
	sort.Strings(synthGroups)

	return json.Marshal(&struct {
		*Alias
		Categories      []string `json:"categories"`
		Languages       []string `json:"languages"`
		Tags            []string `json:"tags"`
		SynthesisGroups []string `json:"synthesis_groups"`
	}{
		Alias:           (*Alias)(w),
		Categories:      categories,
		Languages:       languages,
		Tags:            tags,
		SynthesisGroups: synthGroups,
	})
}

// Dictionary is a collection of words with metadata.
type Dictionary struct {
	Name        string           `json:"name"`
	Language    string           `json:"language,omitempty"`
	Languages   map[string]bool  `json:"-"`
	Words       map[string]*Word `json:"-"` // normalized -> Word
	GeneratedAt string           `json:"generated_at"`
	SourceDicts map[string]bool  `json:"-"`
	WordType    string           `json:"word_type,omitempty"`
}

// NewDictionary creates a new Dictionary.
func NewDictionary(name string) *Dictionary {
	return &Dictionary{
		Name:        name,
		Languages:   make(map[string]bool),
		Words:       make(map[string]*Word),
		GeneratedAt: time.Now().UTC().Format(time.RFC3339),
		SourceDicts: make(map[string]bool),
	}
}

// AddWord adds or merges a word.
func (d *Dictionary) AddWord(word *Word) {
	if existing, ok := d.Words[word.Normalized]; ok {
		for _, source := range word.Sources {
			existing.AddSource(source)
		}
		for tag := range word.Tags {
			existing.Tags[tag] = true
		}
	} else {
		d.Words[word.Normalized] = word
	}

	for lang := range word.Languages {
		d.Languages[lang] = true
	}
	for _, dict := range word.GetSourceDicts() {
		d.SourceDicts[dict] = true
	}
}

// Count returns word count.
func (d *Dictionary) Count() int {
	return len(d.Words)
}

// GetWordsSorted returns sorted list of words.
func (d *Dictionary) GetWordsSorted() []*Word {
	words := make([]*Word, 0, len(d.Words))
	for _, w := range d.Words {
		words = append(words, w)
	}
	sort.Slice(words, func(i, j int) bool {
		return words[i].Normalized < words[j].Normalized
	})
	return words
}

// MarshalJSON implements custom JSON marshaling.
func (d *Dictionary) MarshalJSON() ([]byte, error) {
	languages := make([]string, 0, len(d.Languages))
	for k := range d.Languages {
		languages = append(languages, k)
	}
	sort.Strings(languages)

	sourceDicts := make([]string, 0, len(d.SourceDicts))
	for k := range d.SourceDicts {
		sourceDicts = append(sourceDicts, k)
	}
	sort.Strings(sourceDicts)

	wordsMap := make(map[string]*Word)
	for _, w := range d.GetWordsSorted() {
		wordsMap[w.Normalized] = w
	}

	return json.Marshal(&struct {
		Name        string           `json:"name"`
		Language    string           `json:"language,omitempty"`
		Languages   []string         `json:"languages"`
		WordType    string           `json:"word_type,omitempty"`
		GeneratedAt string           `json:"generated_at"`
		SourceDicts []string         `json:"source_dicts"`
		WordCount   int              `json:"word_count"`
		Words       map[string]*Word `json:"words"`
	}{
		Name:        d.Name,
		Language:    d.Language,
		Languages:   languages,
		WordType:    d.WordType,
		GeneratedAt: d.GeneratedAt,
		SourceDicts: sourceDicts,
		WordCount:   d.Count(),
		Words:       wordsMap,
	})
}

// Save saves dictionary to JSON file.
func (d *Dictionary) Save(filePath string) error {
	dir := filepath.Dir(filePath)
	if err := os.MkdirAll(dir, 0755); err != nil {
		return err
	}

	file, err := os.Create(filePath)
	if err != nil {
		return err
	}
	defer file.Close()

	encoder := json.NewEncoder(file)
	encoder.SetIndent("", "  ")
	return encoder.Encode(d)
}

// WordTypeFromLength returns word type string from length.
func WordTypeFromLength(length int) string {
	if length >= 3 && length <= 10 {
		return string(rune('0'+length/10)) + string(rune('0'+length%10)) + "-c"[1:]
	}
	return ""
}

File: python/ditong/schema.py
"""Word schema and data structures for ditong.

Core concept:
    - Words from multiple sources normalize to the same form
    - Each word tracks its origin sources and metadata
    - Synthesis configs filter words by source tags

Example:
    "care" (EN) + "çare" (TR) → normalized "care"
    Tagged with sources: ["hunspell_en_us", "hunspell_tr"]
"""

from dataclasses import dataclass, field
from datetime import datetime, timezone
from enum import Enum
from pathlib import Path
from typing import Any, Optional
import json


class WordType(Enum):
    """Word length type identifier."""

    C3 = "3-c"
    C4 = "4-c"
    C5 = "5-c"
    C6 = "6-c"
    C7 = "7-c"
    C8 = "8-c"
    C9 = "9-c"
    C10 = "10-c"

    @classmethod
    def from_length(cls, length: int) -> Optional["WordType"]:
        """Get WordType from character length."""
        mapping = {
            3: cls.C3, 4: cls.C4, 5: cls.C5, 6: cls.C6,
            7: cls.C7, 8: cls.C8, 9: cls.C9, 10: cls.C10,
        }
        return mapping.get(length)


@dataclass
class WordSource:
    """Tracks where a word came from."""

    dict_name: str          # e.g., "hunspell_en_us", "urban_dictionary"
    dict_filepath: str      # Full path to source file
    language: str           # e.g., "en", "tr"
    original_form: str      # Original word before normalization
    line_number: Optional[int] = None  # Line in source file (if applicable)
    category: str = "standard"  # e.g., "standard", "urban", "curseword"

    def to_dict(self) -> dict[str, Any]:
        """Convert to dictionary for JSON serialization."""
        return {
            "dict_name": self.dict_name,
            "dict_filepath": self.dict_filepath,
            "language": self.language,
            "original_form": self.original_form,
            "line_number": self.line_number,
            "category": self.category,
        }

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> "WordSource":
        """Create from dictionary."""
        return cls(
            dict_name=data["dict_name"],
            dict_filepath=data["dict_filepath"],
            language=data["language"],
            original_form=data["original_form"],
            line_number=data.get("line_number"),
            category=data.get("category", "standard"),
        )


@dataclass
class Word:
    """A normalized word with full metadata."""

    normalized: str                         # Normalized form (lowercase, ASCII)
    length: int                             # Character count
    word_type: str                          # e.g., "5-c"
    sources: list[WordSource] = field(default_factory=list)
    categories: set[str] = field(default_factory=set)
    languages: set[str] = field(default_factory=set)
    tags: set[str] = field(default_factory=set)
    ipa: Optional[str] = None               # IPA transcription (if computed)
    synthesis_groups: set[str] = field(default_factory=set)

    def __post_init__(self):
        """Derive fields from sources."""
        for source in self.sources:
            self.categories.add(source.category)
            self.languages.add(source.language)

    def add_source(self, source: WordSource) -> None:
        """Add a source and update derived fields."""
        self.sources.append(source)
        self.categories.add(source.category)
        self.languages.add(source.language)

    def get_source_dicts(self) -> list[str]:
        """Get list of source dictionary names."""
        return [s.dict_name for s in self.sources]

    def matches_filter(
        self,
        include_categories: Optional[set[str]] = None,
        exclude_categories: Optional[set[str]] = None,
        include_languages: Optional[set[str]] = None,
        min_length: Optional[int] = None,
        max_length: Optional[int] = None,
    ) -> bool:
        """Check if word matches filter criteria."""
        if min_length and self.length < min_length:
            return False
        if max_length and self.length > max_length:
            return False
        if include_languages and not self.languages & include_languages:
            return False
        if include_categories and not self.categories & include_categories:
            return False
        if exclude_categories and self.categories & exclude_categories:
            return False
        return True

    def to_dict(self) -> dict[str, Any]:
        """Convert to dictionary for JSON serialization."""
        return {
            "normalized": self.normalized,
            "length": self.length,
            "type": self.word_type,
            "sources": [s.to_dict() for s in self.sources],
            "categories": sorted(self.categories),
            "languages": sorted(self.languages),
            "tags": sorted(self.tags),
            "ipa": self.ipa,
            "synthesis_groups": sorted(self.synthesis_groups),
        }

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> "Word":
        """Create from dictionary."""
        word = cls(
            normalized=data["normalized"],
            length=data["length"],
            word_type=data["type"],
            sources=[WordSource.from_dict(s) for s in data.get("sources", [])],
            ipa=data.get("ipa"),
        )
        word.categories = set(data.get("categories", []))
        word.languages = set(data.get("languages", []))
        word.tags = set(data.get("tags", []))
        word.synthesis_groups = set(data.get("synthesis_groups", []))
        return word


@dataclass
class Dictionary:
    """A collection of words with metadata."""

    name: str                               # e.g., "en_5-c", "en_tr_synthesis"
    language: Optional[str] = None          # Single language or None for synthesis
    languages: set[str] = field(default_factory=set)
    words: dict[str, Word] = field(default_factory=dict)  # normalized -> Word
    generated_at: str = field(
        default_factory=lambda: datetime.now(timezone.utc).isoformat()
    )
    source_dicts: set[str] = field(default_factory=set)
    word_type: Optional[str] = None         # e.g., "5-c" for length-specific dicts

    def add_word(self, word: Word) -> None:
        """Add or merge a word."""
        if word.normalized in self.words:
            existing = self.words[word.normalized]
            for source in word.sources:
                existing.add_source(source)
            existing.tags.update(word.tags)
        else:
            self.words[word.normalized] = word

        self.languages.update(word.languages)
        self.source_dicts.update(word.get_source_dicts())

    def get_words_list(self) -> list[Word]:
        """Get sorted list of words."""
        return sorted(self.words.values(), key=lambda w: w.normalized)

    def count(self) -> int:
        """Get word count."""
        return len(self.words)

    def to_dict(self) -> dict[str, Any]:
        """Convert to dictionary for JSON serialization."""
        return {
            "name": self.name,
            "language": self.language,
            "languages": sorted(self.languages),
            "word_type": self.word_type,
            "generated_at": self.generated_at,
            "source_dicts": sorted(self.source_dicts),
            "word_count": self.count(),
            "words": {
                w.normalized: w.to_dict() for w in self.get_words_list()
            },
        }

    def save(self, filepath: Path) -> None:
        """Save dictionary to JSON file."""
        filepath.parent.mkdir(parents=True, exist_ok=True)
        with open(filepath, "w", encoding="utf-8") as f:
            json.dump(self.to_dict(), f, indent=2, ensure_ascii=False)

    @classmethod
    def load(cls, filepath: Path) -> "Dictionary":
        """Load dictionary from JSON file."""
        with open(filepath, "r", encoding="utf-8") as f:
            data = json.load(f)

        d = cls(
            name=data["name"],
            language=data.get("language"),
            word_type=data.get("word_type"),
            generated_at=data.get("generated_at", ""),
        )
        d.languages = set(data.get("languages", []))
        d.source_dicts = set(data.get("source_dicts", []))

        for word_data in data.get("words", {}).values():
            d.add_word(Word.from_dict(word_data))

        return d
